iter/iter.go

```diff
@@ -83,3 +102,55 @@ func (iter Iterator[T]) ForEachIdx(input []T, f func(int, *T)) {
 	}
 	wg.Wait()
 }
+
+// ForEachErr is the same as ForEach except it returns an error.
+// If Iterator.FailFast is true, ForEachErr will stop iterating on the first error.
+// If Iterator.FailFast is false, ForEachErr will return all errors and iterate over the all the input.
+func (iter Iterator[T]) ForEachErr(input []T, f func(*T) error) error {
+	return iter.ForEachIdxErr(input, func(_ int, t *T) error {
+		return f(t)
+	})
+}
+
+// ForEachIdxErr is the same as ForEachIdx except it returns an error.
+// If Iterator.FailFast is true, ForEachIdxErr will stop iterating on the first error.
+// If Iterator.FailFast is false, ForEachIdxErr will return all errors and iterate over the all the input.
+func (iter Iterator[T]) ForEachIdxErr(input []T, f func(int, *T) error) error {
+	if iter.MaxGoroutines == 0 {
+		// iter is a value receiver and is hence safe to mutate
+		iter.MaxGoroutines = defaultMaxGoroutines()
+	}
+
+	numInput := len(input)
+	if iter.MaxGoroutines > numInput {
+		// No more concurrent tasks than the number of input items.
+		iter.MaxGoroutines = numInput
+	}
+
+	var errs error
+	var errsMu sync.Mutex
+	var idx atomic.Int64
+	var failed atomic.Bool
+
+	// Create the task outside the loop to avoid extra closure allocations.
+	task := func() {
+		i := int(idx.Add(1) - 1)
+		for ; i < numInput && !failed.Load(); i = int(idx.Add(1) - 1) {
+			if err := f(i, &input[i]); err != nil {
+				errsMu.Lock()
+				errs = multierror.Join(errs, err)
+				errsMu.Unlock()
+
+				failed.Store(iter.FailFast)
+			}
+		}
+	}
```

Comment:
It looks to me like FailFast does not guarantee returning only the error that caused the failure. One task can check if failed was set, error, and append its error while another task concurrently failed and appends an error.

I think this counts as surprising behavior given that the rest of the library guarantees that only the first error is retained when WithFirstError() is used. 

I think we can fix this by checking the old value of `failed`:

```suggestion
	task := func() {
		i := int(idx.Add(1) - 1)
		for ; i < numInput && !failed.Load(); i = int(idx.Add(1) - 1) {
			if err := f(i, &input[i]); err != nil {
				if alreadyFailedFast := failed.Swap(iter.FailFast); !alreadyFailedFast {
					errsMu.Lock()
					errs = multierror.Join(errs, err)
					errsMu.Unlock()
				}
			}
		}
	}
```

Reply:
fixed 👍